Mathlib/MeasureTheory/Measure/Tight.lean

```diff
@@ -0,0 +1,461 @@
+  use u, hu
+  intro j
+  obtain ⟨N, hN⟩ := cover (ε/(2^(j+1))) (hεj_pos j) (1/(j+1)) (Nat.one_div_pos_of_nat)
+  use N
+  rw [measure_compl (by measurability) (measure_ne_top μ _)]
+  exact tsub_le_iff_tsub_le.mp hN
+
+/-- A measure `μ` is separable if there is a separable set `S` such that `μ S = μ Set.univ`. -/
+ def IsSeparable [TopologicalSpace α] (μ : Measure α) : Prop :=
+  ∃ S : Set α, TopologicalSpace.IsSeparable S ∧ μ S = μ Set.univ
+
+namespace IsSeparable
+
+lemma of_separableSpace [TopologicalSpace α] [TopologicalSpace.SeparableSpace α] :
+    IsSeparable μ := ⟨Set.univ, TopologicalSpace.isSeparable_univ_iff.mpr ‹_›, rfl⟩
+
+end IsSeparable
+
+section IsPretight
+
+variable [UniformSpace α]
+
+/-- A measure `μ` is pre-tight if for all `0 < ε`, there exists `K` totally bounded such that
+  `μ Kᶜ ≤ ε`. -/
+def IsPretight (μ : Measure α) : Prop :=
+  ∀ ε : ℝ≥0∞, 0 < ε → ∃ K : Set α, TotallyBounded K ∧ μ Kᶜ ≤ ε
+
+namespace IsPretight
+
+lemma of_totallyBounded (h : TotallyBounded (Set.univ : Set α)) : IsPretight μ := by
+  intro ε hε
+  use Set.univ, h
+  rw [Set.compl_univ, measure_empty]
+  exact hε.le
+
+lemma exists_totallyBounded_nat (h : IsPretight μ) (n : ℕ):
+    ∃ K : Set α, TotallyBounded K ∧ μ Kᶜ ≤ 1/n := by
+  apply h
+  simp only [one_div, ENNReal.inv_pos, ne_eq, ENNReal.natCast_ne_top, not_false_eq_true]
+
+/-- It suffices to check totally boundedness along countably many `ε`. -/
+lemma of_totallyBounded_nat (h : ∀ n : ℕ, ∃ K : Set α, TotallyBounded K ∧ μ Kᶜ ≤ 1/n) :
+    IsPretight μ := by
+  intro ε hε
+  obtain ⟨n, hn⟩ := ENNReal.exists_inv_nat_lt hε.ne'
+  obtain ⟨K, hK, hKe⟩ := h n
+  refine ⟨K, hK, ?_⟩
+  apply le_trans hKe (le_trans ?_ hn.le)
+  rw [one_div, ENNReal.inv_le_inv]
+
+lemma totallyBounded_nat_iff : IsPretight μ ↔ ∀ n : ℕ, ∃ K : Set α, TotallyBounded K ∧ μ Kᶜ ≤ 1/n :=
+  ⟨exists_totallyBounded_nat, of_totallyBounded_nat⟩
+
+/-- If a measure `μ` is pretight, we can cover `μ`-almost all of the space by a countable sequence
+of totally bounded sets. -/
+lemma exists_countable_totallyBounded_union (h : IsPretight μ):
+    ∃ K : ℕ → Set α, (∀ n, TotallyBounded (K n)) ∧ μ (⋃ n, K n) = μ Set.univ := by
+  choose! K hKa hKb using h.exists_totallyBounded_nat
+  use K, hKa
+  rw [← Set.iUnion_accumulate, measure_congr]
+  rw [ae_eq_univ, Set.compl_iUnion, ← le_zero_iff]
+  refine le_of_forall_lt' (fun ε hε ↦ ?_)
+  obtain ⟨n, hn⟩ := ENNReal.exists_inv_nat_lt hε.ne.symm
+  rw [← one_div] at hn
+  have : μ ((Set.Accumulate K n)ᶜ) ≤ μ ((K n)ᶜ) := by
+    apply measure_mono
+    rw [Set.compl_subset_compl]
+    exact Set.subset_accumulate
+  exact lt_of_le_of_lt (measure_mono <| Set.iInter_subset _ n)
+    (lt_of_le_of_lt this (lt_of_le_of_lt (hKb n) hn))
+
+/-- Every pretight measure on a countably generated uniform space is separable. -/
+lemma to_isSeparable_on_countably_generated_uniformSpace [IsCountablyGenerated (𝓤 α)]
+    (h : IsPretight μ) : IsSeparable μ := by
```

Comment:
When dealing with separablility, you assume that the uniformity filter is countably generated, and in that case separable is equivalent to secound-countable (see [UniformSpace.secondCountable_of_separable](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/Cauchy.html#UniformSpace.secondCountable_of_separable)). This makes me think that the right notion is maybe secound-countable? Again tell me if I am wrong.

Reply:
I'm not restricting to countably generated uniformity filters in general, but I think this condition is needed to get the implication pretight -> separable. I think that the key ingredient of IsSeparable is that you get a separable support, so I would prefer to use this definition. We could also introduce some IsSecondCountablySupported definition, but I don't think that will have a lot of usage (I've never seen this at least).

Reply:
Well it seems to me that whenever you deal with separability in theorems, you either assume that the uniformity filter is countably generated or the space is pseudo metric, so in each case second-countable does not lose generality. But if you are convinced that separability adds something, feel free to resolve conversation.

Reply:
I agree that second-countable does not lose generality. What I meant is that the definition in the literature seems to be in terms of separable support, rather than in terms of second-countable support, so I'm not sure if using a different definition is efficient here?

Reply:
I see. I do not have the answer, so maybe it's simpler if you leave it as it is.